### ISSUE TITLE
Unexpected "permission denied" while running "fastly compute serve"

### ISSUE THREAD

**Issue by dnechay** [fastly/cli#364]
Hey
Playing around with your nice Compute@Edge engine and faced the next issue, any help is appreciated.

### Environment
```
NAME="Linux Mint"
VERSION="20.2 (Uma)"
ID=linuxmint
ID_LIKE=ubuntu
PRETTY_NAME="Linux Mint 20.2"
VERSION_ID="20.2"
HOME_URL="https://www.linuxmint.com/"
SUPPORT_URL="https://forums.linuxmint.com/"
BUG_REPORT_URL="http://linuxmint-troubleshooting-guide.readthedocs.io/en/latest/"
PRIVACY_POLICY_URL="https://www.linuxmint.com/"
VERSION_CODENAME=uma
UBUNTU_CODENAME=focal
```

### Fastly CLI Version

Fastly CLI version 0.36.0 (dd0b7d9)
Built with go version go1.16.6 linux/amd64

### What happened

1. `fastly compute init` (for javascript and from https://github.com/fastly/compute-starter-kit-javascript-default)
2. `fastly compute serve`

**Expected behavior**
Get the local dev env running

**Actual Behavior**
Successfully built but weren't able to start `viceroy`
![Screenshot from 2021-08-06 12-31-29](https://user-images.githubusercontent.com/15381780/128490563-250f65fd-dc1e-4b1c-a99a-fc2931f17cba.png)

**Steps taken**
- tried to reinstall CLI and run commands one more time - no results
- tried to run `serve` command with `sudo`, but locally I use `nvm` to manage Node.js installations, so under `sudo` there is no `npm` and `fastly` isn't able to build package; also IMO `serve` command should not require any superuser permission

**Comment by kailan** [comment]
Hey @dnechay, thanks for reporting!

There is a known workaround for this: `chmod +x ~/.config/fastly/viceroy`

I'll leave the issue open to track this until a proper fix next week 👍

**Comment by dnechay** [comment]
> chmod +x ~/.config/fastly/viceroy

Thanks @kailan , this workaround works for me!

**Comment by Integralist** [comment]
Just a quick update to state that it looks like we already attempt to set the permissions on the downloaded binary:

https://github.com/fastly/cli/blob/3209162a6f8e689766d8197e6bcc4fb785ce8dfa/pkg/commands/update/versioner.go#L190-L193

I'm seeing if I can replicate this issue, but the issue looks to affect Linux users only (from what I can see from other reports) and I'm on macOS so this might affect my ability to reproduce accurately (I'll be debugging using Docker).

**Comment by Integralist** [comment]
So far I've been unable to replicate using either:

- `docker run -it ubuntu`
- `docker run -it linuxmintd/mint20.2-amd64`

```bash
apt-get update
apt-get install -y curl git
curl -fsSL https://deb.nodesource.com/setup_16.x | bash -
apt-get install -y nodejs
cd tmp && mkdir testproject && cd testproject
curl -L -o fastly_0.36.0_linux_amd64.deb https://github.com/fastly/cli/releases/download/v0.36.0/fastly_0.36.0_linux_amd64.deb
apt install ./fastly_*_linux_amd64.deb
rm fastly_0.36.0_linux_amd64.deb

fastly version
Fastly CLI version 0.36.0 (dd0b7d9)
Built with go version go1.16.6 linux/amd64
```

On both OS's I'm seeing the Viceroy binary having the relevant executable permissions applied.

**Comment by dnechay** [comment]
@Integralist by default docker containers run as the root user, so this is why it can be ok there.
I assume if you add some user there or use actual linux distro it will be possible to reproduce the issue

**Comment by Integralist** [comment]
Thanks @dnechay I'll try spinning up a user account inside the docker container to see if that helps with reproducing the issue.

**Comment by Integralist** [comment]
@dnechay are you able to download the binary directly from its repo: https://github.com/fastly/Viceroy/releases/tag/v0.2.2 and let me know what the execution permissions are. For example, I just checked the macOS release directly and it has the correct executable permissions so I'm curious to know if the linux one is the same, and if it shows up as having executable permissions for you.

Thanks!

**Comment by Integralist** [comment]
One of my colleagues created an example Dockerfile to try and replicate the issue, while avoiding root (i.e. adding a new user and installing via that user) but this also didn't yield a reproducible example...

```Dockerfile
FROM linuxmintd/mint20.2-amd64

ENV DEBIAN_FRONTEND=noninteractive

RUN \
      apt-get update -y && \
      apt-get install -y sudo && \
      useradd fastly -G sudo --create-home && \
      echo '%sudo ALL=(ALL) NOPASSWD:ALL' >> /etc/sudoers && \
      curl -fsSL https://deb.nodesource.com/setup_16.x | bash -E - && \
      apt-get install -y curl git nodejs npm

USER fastly
WORKDIR /home/fastly/testapp

RUN \
      curl -L -o /tmp/fastly_0.36.0_linux_amd64.deb https://github.com/fastly/cli/releases/download/v0.36.0/fastly_0.36.0_linux_amd64.deb && \
      sudo apt install -y /tmp/fastly_*_linux_amd64.deb 
```

Next steps is still to get a reproducible example. I'll try to get access to an actual Linux instance for testing.

**Comment by dnechay** [comment]
> @dnechay are you able to download the binary directly from its repo: https://github.com/fastly/Viceroy/releases/tag/v0.2.2 and let me know what the execution permissions are. For example, I just checked the macOS release directly and it has the correct executable permissions so I'm curious to know if the linux one is the same, and if it shows up as having executable permissions for you.
> 
> Thanks!

Downloaded this, unpacked, permissions are `-rwxr-xr-x`.
If I can provide any additional info - please let me know.

**Comment by Integralist** [comment]
@dnechay I have what I hope is a fix to your issue. It would be really helpful if you could validate the changes in https://github.com/fastly/cli/pull/368 resolve your issue.

- `git clone git@github.com:fastly/cli.git`
- `cd cli`
- `git checkout integralist/viceroy-perms`
- `make build`
- `cd /path/to/your/project`
- `/path/to/cli/repo/fastly compute serve`

Thanks.

**Comment by dnechay** [comment]
@Integralist built it with `go1.16.7 linux/amd64` and can verify that issue is resolved in #368 

Thanks for your assistance!

**Comment by Integralist** [comment]
Great news! Thank you for taking the time to validate this change for me. I'll aim to get a new CLI release out ASAP.
